### PR TITLE
Mamba & RecurrentGemma: enable strict signature

### LINKED ISSUE #31540: Batch Generation giving different output when using batch size > 1 or when using padding in MambaForCausalLM
### System Info

- `transformers` version: 4.41.2
- Platform: Linux-6.5.0-41-generic-x86_64-with-glibc2.35
- Python version: 3.10.8
- Huggingface_hub version: 0.23.3
- Safetensors version: 0.4.3
- Accelerate version: 0.26.0
- Accelerate config: not found
- PyTorch version (GPU?): 2.0.0+cu117 (True)
- Tensorflow version (GPU?): 2.9.0 (True)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: Yes
- Using distributed or parallel set-up in script?: No

### Who can help?

@ArthurZucker @gante

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

I have trained a MambaForCausalLM model on a custom dataset.
I am using the following code to generate next token in eval mode - 
```
model = MambaForCausalLM.from_pretrained("mamba_custom")
tokenizer = PreTrainedTokenizerFast.from_pretrained("mamba_tokenizer_custom")
tokenizer.padding_side = 'left'

inputs_rl = tokenizer(sentence, padding="max_length", truncation=True, max_length=100, return_tensors="pt").to("cuda:0")
with torch.no_grad():
    outputs = model(inputs_rl["input_ids"], attention_mask=inputs_rl["attention_mask"])
```

### Expected behavior

The tokenizer pads the input towards the left side. When I change the argument padding="max_length" to generate inputs without padding, I get different tokens as prediction.

Using model.generate gives the same issue as well.

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

Fixes #31540

Mamba accepts `**kwargs`, and thus `attention_mask` can be passed. Many users thus assume it behaves just like other models and can support left-padding.

RecurrentGemma also accept `**kwargs`, but simply not to crash `generate`.

This PR enables a strict signature on Mamba and RecurrentGemma.

### CODE REVIEW

File: src/transformers/models/mamba/modeling_mamba.py [hunk 1]
```diff
@@ -545,7 +545,6 @@ def forward(
         use_cache: Optional[bool] = None,
         output_hidden_states: Optional[bool] = None,
         return_dict: Optional[bool] = None,
-        **kwargs,  # `attention_mask` is passed by the tokenizer and we don't want it
```
Comment (gante) [inline src/transformers/models/mamba/modeling_mamba.py:548]: alternatively, we can accept `attention_mask` and raise an exception when it is not `None`

File: src/transformers/generation/utils.py [hunk 2]
```diff
@@ -2642,13 +2642,12 @@ def _sample(
             # prepare model inputs
             model_inputs = self.prepare_inputs_for_generation(input_ids, **model_kwargs)
 
+            # prepare variable output controls (note: some models won't accept all output controls)
+            model_inputs.update({"output_attentions": output_attentions} if output_attentions else {})
+            model_inputs.update({"output_hidden_states": output_hidden_states} if output_hidden_states else {})
```
Comment (ArthurZucker) [inline src/transformers/generation/utils.py:2647]: yesssss I think I have a PR open where I dod this! Finally!

**Review by ArthurZucker — APPROVED** [review]
Let's googoogogogogo 🚀

File: src/transformers/models/mamba/modeling_mamba.py (hunk 1 above)
Comment (amyeroberts) [inline src/transformers/models/mamba/modeling_mamba.py:548]: Removing this will break FDSP :( See #31161
